### ISSUE TITLE
Python references do not keep engine objects alive

### ISSUE THREAD

**Issue by dfb** [20tab/UnrealEnginePython#656]
We've been running into some cases where the engine crashes and we've found that it's because the objects we're referencing in Python have been garbage collected out from under us - the Python object is still valid and has a non-zero reference count, but the UObject it wraps is no longer valid.

If we enable UEPY_MEMORY_DEBUG, then we can see the following line in PythonHousekeeper.h's PyUObjectGC method get called, because Tracker.Owner is no longer valid:

```py
UE_LOG(LogPython, Warning, TEXT("Removing UObject at %p (refcnt: %d)"), Object, Tracker.PyUO...
```

The problem seems to be that when an engine object is wrapped in a ue_PyUObject, it is registered with the housekeeper but that object is not added to the PythonTrackedObjects list. As a result, the PythonHousekeeper doesn't keep the UObject alive even though there is some Python variable somewhere holding a reference to it. In some cases the result is an error message and in other cases UE4 just crashes.

Here is an example of how to reproduce it:
```py
import unreal_engine as ue
from unreal_engine.classes import Character
from unreal_engine.enums import EPixelFormat

class CrashChar(Character):
    def __init__(self):
        self.lastPrint = 0.0

    def ReceiveBeginPlay(self):
        self.texture = ue.create_transient_texture(1024, 1024, EPixelFormat.PF_R8G8B8A8)

    def ReceiveTick(self, deltaSeconds:float):
        self.lastPrint += deltaSeconds
        if self.lastPrint > 2:
            self.lastPrint = 0
            print('B:',deltaSeconds, self.texture.is_valid())

ue.get_editor_world().actor_spawn(CrashChar)
```

If you import this module and then run PIE and let it sit there for a few minutes, it will successfully log messages but after awhile (presumable after the GC runs), it will start printing this error:
```py
Exception: PyUObject is in invalid state
```

If something else in the engine has a reference to the UObject, then things will continue to work. For example, in the plotter tutorial, the above line that creates the transient texture is followed by a call to ```StaticMeshComponent.OverrideMaterials[0].set_material_texture_parameter```, and so the code never crashes because it's the static mesh comp that is keeping the texture alive.

The net result is that from the Python side you can't look at the code and know for sure if you're ok or not, and so it's very, very easy to create scenarios in which something works for awhile then crashes, or scenarios in which changing code in one place causes a failure in something seemingly unrelated.

It seems that the ue_PyUObject should keep the object it is wrapping alive - once there is no Python code accessing it, the ue_PyUObject will be destroyed, and its entry in the housekeeper removed.

I'm going to work on a patch to do this, but I thought I'd submit this issue to see if there is anything I'm misunderstanding.

**Comment by dfb** [comment]
In case anyone is interested, I've [implemented a potential fix](https://github.com/dfb/UnrealEnginePython/commit/3111c3cd37b75c7f6e36aee94dab42d0b7755d8f) that took care of the specific problems I was seeing; I'm now testing it more broadly to see if it needs more work.

**Comment by dfb** [comment]
I've read through [MemoryManagerment](https://github.com/20tab/UnrealEnginePython/blob/master/docs/MemoryManagement.md) and it's sounding like this is the intended behavior, although this really feels like it can make the Python code extremely brittle and error-prone.

The doc says, "This [i.e., the current design] is a pretty complex choice aimed at improving performance and avoiding too much competition between the two GCs." Does this mean that an earlier version didn't do this? I'd love to learn more about it if so, and what the findings were.

In particular, I'd like to know if the PythonHouseKeeper was originally like this or could instead operate like this:
- any engine object passed to Python code is wrapped in a ue_PyObject (just like it is today)
- if a wrapper object already exists for an engine obj, the wrapper object is reused (just like it is today)
- anytime a wrapper object is created, its engine object is added to PythonHouseKeeper's list of tracked objects so the engine object isn't GC'd
- anytime a wrapper object is dealloc'd, its engine object is removed from PythonHouseKeeper's list of tracked objects

That's it. PythonHouseKeeper wouldn't have Register functionality separate from Track (ditto for Unregister and Untrack). It seems like the performance would be more or less the same, but the implementation would be less complex and it would avoid this problem of Python objects becoming invalid because an engine object is GC'd out from under it.

Conceptually, this seems to make the most sense and it fits in with both UE's and Python's garbage collection models (and that's really the big hangup here: normally in Python code, if you save a reference to an object, you're keeping it alive, but that's not how it always works today in UEP, and when it doesn't work that way).

**Comment by abeyondc** [comment]
yes, It's a problem about both UE's and Python's garbage collection 

PythonHouseKeeper should owned the UEPythonObject when init,because destructor function will always Py_XDECREF(py_uobject)

I fixed by add a func  FUnrealEnginePythonHouseKeeper::Get()->ForceSetPythonOwned(this);
on the UEPythonObject PreInitializeComponents.

if it's owned GC of PythonHouseKeeper  will not Py_XDECREF(py_uobject)

**Comment by dfb** [comment]
My earlier attempt at fixing this was insufficient; I've taken another stab at it here:

https://github.com/dfb/UnrealEnginePython/tree/pr_py_obj_keepalive

Basically: anytime an engine object is passed to Python code, the engine object is added to the housekeeper to ensure the object remains alive. Once no Python code is accessing the object anymore, it is removed from the housekeeper.

**Comment by dfb** [comment]
FWIW my fork is starting to diverge because we've also implemented a different approach to subclassing in Python engine objects, that fixes some of the problems with the subclassing approach of UEPy. Between the different subclassing and this issue, we're introducing a lot of changes but things are generally working quite well for us in production (i.e. using Python at runtime).
